akka-remote/src/main/scala/akka/remote/artery/ArteryTcpTransport.scala

```diff
@@ -0,0 +1,365 @@
+      Sink.fromGraph(GraphDSL.create() { implicit b ⇒
+        import GraphDSL.Implicits._
+        val partition = b.add(Partition[EnvelopeBuffer](3, env ⇒ {
+          env.byteBuffer.get(EnvelopeBuffer.StreamIdOffset).toInt match {
+            case `ordinaryStreamId` ⇒ 1
+            case `controlStreamId`  ⇒ 0
+            case `largeStreamId`    ⇒ 2
+          }
+        }))
+        partition.out(0) ~> controlStream
+        partition.out(1) ~> ordinaryMessagesStream
+        partition.out(2) ~> largeMessagesStream
+        SinkShape(partition.in)
+      })
+
+    val maxFrameSize =
+      if (largeMessageChannelEnabled) settings.Advanced.MaximumLargeFrameSize
+      else settings.Advanced.MaximumFrameSize
+
+    // FIXME error handling in case something fails in this part
+
+    val inboundConnectionFlow = Flow[ByteString]
+      .via(inboundConnectionsKillSwitch.flow)
+      .via(Framing.lengthField(fieldLength = 4, fieldOffset = EnvelopeBuffer.FrameLengthOffset,
+        maxFrameSize, byteOrder = ByteOrder.LITTLE_ENDIAN))
+      .map { frame ⇒
+        val buffer = ByteBuffer.wrap(frame.toArray)
+        buffer.order(ByteOrder.LITTLE_ENDIAN)
+        new EnvelopeBuffer(buffer)
+      }
+      .alsoTo(inboundStream)
+      .filter(_ ⇒ false) // don't send back anything in this TCP socket
+      .map(_ ⇒ ByteString.empty) // make it a Flow[ByteString] again
+
+    val host = localAddress.address.host.get
+    val port = localAddress.address.port.get
+
+    serverBinding =
+      Some(Tcp().bind(
+        interface = localAddress.address.host.get,
+        port = localAddress.address.port.get,
+        halfClose = false)
+        .to(Sink.foreach { connection ⇒
+          if (tlsEnabled) {
+            val rhs: Flow[SslTlsInbound, SslTlsOutbound, Any] =
+              Flow[SslTlsInbound]
+                .collect {
+                  case SessionBytes(_, bytes) ⇒ bytes
+                }
+                .via(inboundConnectionFlow)
+                .map(SendBytes.apply)
+
+            connection.handleWith(tls(role = Server).reversed join rhs)
+          } else
+            connection.handleWith(inboundConnectionFlow)
+        })
+        .run()
+        .recoverWith {
+          case e ⇒ Future.failed(new RemoteTransportException(
+            s"Failed to bind TCP to [${localAddress.address.host.get}:${localAddress.address.port.get}] due to: " +
+              e.getMessage, e))
+        }(ExecutionContexts.sameThreadExecutionContext)
+      )
+
+    Await.result(serverBinding.get, settings.Bind.BindTimeout)
+
+    implicit val ec = materializer.executionContext
+    val completed = Future.firstCompletedOf(
+      List(controlStreamCompleted, ordinaryMessagesStreamCompleted, largeMessagesStreamCompleted))
+    val restart = () ⇒ {
+      inboundConnectionsKillSwitch.shutdown()
+      inboundConnectionsKillSwitch = KillSwitches.shared("inboundConnectionsKillSwitch")
+      unbind().onComplete { _ ⇒
```

Comment:
Do we need to unbind or can we just kill or the existing connections? Or is that too complicated?

Reply:
Probably not needed, but I found it easier to start all over.
I'll add a FIXME and look into it one more time.

Note that failures in the inbound streams should be very rare (I can't think of any case other than bugs). I clarified that in a later commit.